File: frontend/src/types/incident.ts
/** Domain models for Ajali! incident operations (Sprint 1 mock + Sprint 2 Flask-ready). */

export type IncidentStatus =
  | "PENDING"
  | "VERIFIED"
  | "IN_PROGRESS"
  | "RESOLVED"
  | "CLOSED"

/** How quickly the report needs administrative attention. */
export type IncidentUrgency = "LOW" | "MEDIUM" | "HIGH" | "CRITICAL"

/** How serious the actual incident is. */
export type IncidentSeverity = "MINOR" | "MODERATE" | "MAJOR" | "CRITICAL"

export type IncidentType =
  | "accident"
  | "fire"
  | "medical"
  | "crime"
  | "disaster"

export type NotificationChannel = "IN_APP" | "EMAIL" | "SMS"

export type PreferredContactMethod = "PHONE" | "EMAIL" | "OTHER"

export type VerificationStatus = "PENDING" | "VERIFIED" | "FAILED"

export type VerificationMethod = "PHONE" | "EMAIL" | "OTHER"

export type CloseReasonCode =
  | "FALSE_REPORT"
  | "DUPLICATE"
  | "UNABLE_TO_VERIFY"
  | "INSUFFICIENT_INFORMATION"
  | "OTHER"

export type DepartmentType =
  | "POLICE"
  | "FIRE"
  | "HOSPITAL"
  | "AMBULANCE"
  | "DISASTER_RESPONSE"
  | "OTHER"

export type HandoffStatus =
  | "PENDING"
  | "ACKNOWLEDGED"
  | "IN_PROGRESS"
  | "COMPLETED"
  | "CANCELLED"

export type ResolutionOutcome =
  | "RESOLVED"
  | "ASSISTANCE_PROVIDED"
  | "REFERRED"
  | "UNABLE_TO_ASSIST"
  | "OTHER"

export type Incident = {
  id: string
  reference: string
  title: string
  description: string
  type: IncidentType
  urgency: IncidentUrgency
  severity: IncidentSeverity
  status: IncidentStatus
  location: string
  lat: number | null
  lng: number | null
  userId: string
  reporterName?: string
  reporterEmail?: string
  reporterPhone?: string
  preferredContactMethod?: PreferredContactMethod
  /** Latest verification record id when present */
  verificationId?: string
  closeReasonCode?: CloseReasonCode
  resolutionSummary?: string
  resolutionNotes?: string
  resolutionOutcome?: ResolutionOutcome
  resolvedById?: string
  resolvedByName?: string
  resolvedAt?: string
  archived: boolean
  archiveReason?: string
  createdAt: string
  updatedAt: string
}

export type ReporterVerification = {
  id: string
  incidentId: string
  status: VerificationStatus
  method?: VerificationMethod
  notes?: string
  verifiedById?: string
  verifiedByName?: string
  verifiedAt?: string
  createdAt: string
  updatedAt: string
}

export type Department = {
  id: string
  name: string
  type: DepartmentType
  description?: string
  phone?: string
  email?: string
  location?: string
  active: boolean
  createdAt: string
  updatedAt: string
}

export type DepartmentHandoff = {
  id: string
  incidentId: string
  departmentId: string
  initiatedById: string
  initiatedByName: string
  status: HandoffStatus
  notes?: string
  handedOffAt: string
  acknowledgedAt?: string
  completedAt?: string
  updatedAt: string
}

export type IncidentMedia = {
  id: string
  incidentId: string
  kind: "image" | "video"
  url: string
  name: string
  createdAt: string
}

export type IncidentNote = {
  id: string
  incidentId: string
  authorId: string
  authorName: string
  body: string
  createdAt: string
}

export type StatusHistory = {
  id: string
  incidentId: string
  fromStatus: IncidentStatus | null
  toStatus: IncidentStatus
  actorId: string
  actorName: string
  reason?: string
  createdAt: string
}

export type AuditLog = {
  id: string
  incidentId?: string
  /** Denormalized for UI — Flask can join or store */
  incidentReference?: string
  actorId: string
  actorName: string
  action: AuditAction | string
  previousValue?: string
  newValue?: string
  reason?: string
  metadata?: string
  createdAt: string
}

export type AppNotification = {
  id: string
  incidentId?: string
  type: NotificationEventType | string
  channel: NotificationChannel
  title: string
  body: string
  read: boolean
  createdAt: string
}

/** List DTO — includes latest verification without N+1 */
export type IncidentListItem = Incident & {
  verificationStatus: VerificationStatus
}

export type IncidentListQuery = {
  userId?: string
  status?: IncidentStatus
  /** When set, matches any of these statuses (ignored if `status` is set). */
  statusIn?: IncidentStatus[]
  urgency?: IncidentUrgency
  severity?: IncidentSeverity
  type?: IncidentType
  departmentId?: string
  verificationStatus?: VerificationStatus
  location?: string
  search?: string
  dateFrom?: string
  dateTo?: string
  sort?: "urgency" | "newest"
  includeArchived?: boolean
}

export type AuditAction =
  | "REPORT_CREATED"
  | "REPORT_UPDATED"
  | "REPORT_VERIFIED"
  | "REPORT_CLOSED"
  | "RESPONSE_STARTED"
  | "DEPARTMENT_ASSIGNED"
  | "DEPARTMENT_HANDOFF_UPDATED"
  | "DEPARTMENT_CREATED"
  | "DEPARTMENT_UPDATED"
  | "INCIDENT_RESOLVED"
  | "INCIDENT_REOPENED"
  | "INCIDENT_ARCHIVED"
  | "MEDIA_ADDED"
  | "MEDIA_REMOVED"
  | "NOTE_ADDED"
  | "CITIZEN_NOTIFIED"
  | "URGENCY_UPDATED"
  | "SEVERITY_UPDATED"

export type NotificationEventType =
  | "REPORT_RECEIVED"
  | "CRITICAL_REPORT_RECEIVED"
  | "REPORT_VERIFIED"
  | "REPORT_CLOSED"
  | "RESPONSE_STARTED"
  | "DEPARTMENT_ASSIGNED"
  | "INCIDENT_RESOLVED"
  | "INCIDENT_ARCHIVED"
  | "CITIZEN_STATUS_NOTIFY"
  | "CRITICAL_INCIDENT"
  | "STATUS_IN_PROGRESS"

export type DashboardStats = {
  total: number
  pending: number
  verified: number
  inProgress: number
  resolved: number
  closed: number
  /** Critical urgency (admin attention) */
  criticalUrgency: number
  /** Critical severity (incident seriousness) */
  criticalSeverity: number
  awaitingVerification: number
  awaitingResponse: number
  awaitingHandoffAck: number
  today: number
}

export const STATUS_TRANSITIONS: Record<IncidentStatus, IncidentStatus[]> = {
  PENDING: ["VERIFIED", "CLOSED"],
  VERIFIED: ["IN_PROGRESS", "CLOSED"],
  IN_PROGRESS: ["RESOLVED"],
  RESOLVED: ["IN_PROGRESS"],
  CLOSED: ["PENDING"],
}

export const URGENCY_RANK: Record<IncidentUrgency, number> = {
  CRITICAL: 0,
  HIGH: 1,
  MEDIUM: 2,
  LOW: 3,
}

export function canTransition(
  from: IncidentStatus,
  to: IncidentStatus
): boolean {
  return STATUS_TRANSITIONS[from].includes(to)
}

export function statusLabel(status: IncidentStatus): string {
  const labels: Record<IncidentStatus, string> = {
    PENDING: "Pending",
    VERIFIED: "Verified",
    IN_PROGRESS: "In Progress",
    RESOLVED: "Resolved",
    CLOSED: "Closed",
  }
  return labels[status]
}

export function urgencyLabel(urgency: IncidentUrgency): string {
  const labels: Record<IncidentUrgency, string> = {
    LOW: "Low",
    MEDIUM: "Medium",
    HIGH: "High",
    CRITICAL: "Critical",
  }
  return labels[urgency]
}

export function severityLabel(severity: IncidentSeverity): string {
  const labels: Record<IncidentSeverity, string> = {
    MINOR: "Minor",
    MODERATE: "Moderate",
    MAJOR: "Major",
    CRITICAL: "Critical",
  }
  return labels[severity]
}

export function typeLabel(type: IncidentType): string {
  const labels: Record<IncidentType, string> = {
    accident: "Accident",
    fire: "Fire",
    medical: "Medical",
    crime: "Crime",
    disaster: "Disaster",
  }
  return labels[type]
}

export function departmentTypeLabel(type: DepartmentType): string {
  const labels: Record<DepartmentType, string> = {
    POLICE: "Police",
    FIRE: "Fire & Rescue",
    HOSPITAL: "Hospital",
    AMBULANCE: "Ambulance",
    DISASTER_RESPONSE: "Disaster Response",
    OTHER: "Other",
  }
  return labels[type]
}

export function verificationStatusLabel(status: VerificationStatus): string {
  const labels: Record<VerificationStatus, string> = {
    PENDING: "Not verified",
    VERIFIED: "Verified",
    FAILED: "Failed",
  }
  return labels[status]
}

export function closeReasonLabel(code: CloseReasonCode): string {
  const labels: Record<CloseReasonCode, string> = {
    FALSE_REPORT: "False report",
    DUPLICATE: "Duplicate",
    UNABLE_TO_VERIFY: "Unable to verify",
    INSUFFICIENT_INFORMATION: "Insufficient information",
    OTHER: "Other",
  }
  return labels[code]
}

export function handoffStatusLabel(status: HandoffStatus): string {
  const labels: Record<HandoffStatus, string> = {
    PENDING: "Pending",
    ACKNOWLEDGED: "Acknowledged",
    IN_PROGRESS: "In progress",
    COMPLETED: "Completed",
    CANCELLED: "Cancelled",
  }
  return labels[status]
}

export function resolutionOutcomeLabel(outcome: ResolutionOutcome): string {
  const labels: Record<ResolutionOutcome, string> = {
    RESOLVED: "Resolved",
    ASSISTANCE_PROVIDED: "Assistance provided",
    REFERRED: "Referred",
    UNABLE_TO_ASSIST: "Unable to assist",
    OTHER: "Other",
  }
  return labels[outcome]
}

/** Compare urgency first (critical first), then newest createdAt. */
export function compareByUrgencyThenNewest(a: Incident, b: Incident): number {
  const urg = URGENCY_RANK[a.urgency] - URGENCY_RANK[b.urgency]
  if (urg !== 0) return urg
  return b.createdAt.localeCompare(a.createdAt)
}

File: backend/app/schemas/incident.py
from marshmallow import Schema, fields, validate


INCIDENT_TYPES = ["accident", "fire", "medical", "crime", "disaster"]
URGENCIES = ["LOW", "MEDIUM", "HIGH", "CRITICAL"]
SEVERITIES = ["MINOR", "MODERATE", "MAJOR", "CRITICAL"]
CONTACT_METHODS = ["PHONE", "EMAIL", "OTHER"]
STATUSES = ["PENDING", "VERIFIED", "IN_PROGRESS", "RESOLVED", "CLOSED"]


class IncidentSchema(Schema):
    id = fields.String(required=True)
    reference = fields.String(required=True)
    title = fields.String(required=True)
    description = fields.String(required=True)
    type = fields.String(required=True)
    urgency = fields.String(required=True)
    severity = fields.String(required=True)
    status = fields.String(required=True)
    location = fields.String(required=True)
    lat = fields.Float(allow_none=True)
    lng = fields.Float(allow_none=True)
    userId = fields.String(required=True)
    reporterName = fields.String(allow_none=True)
    reporterEmail = fields.String(allow_none=True)
    reporterPhone = fields.String(allow_none=True)
    preferredContactMethod = fields.String(allow_none=True)
    verificationId = fields.String(allow_none=True)
    closeReasonCode = fields.String(allow_none=True)
    resolutionSummary = fields.String(allow_none=True)
    resolutionNotes = fields.String(allow_none=True)
    resolutionOutcome = fields.String(allow_none=True)
    resolvedById = fields.String(allow_none=True)
    resolvedByName = fields.String(allow_none=True)
    resolvedAt = fields.String(allow_none=True)
    archived = fields.Boolean(required=True)
    archiveReason = fields.String(allow_none=True)
    createdAt = fields.String(required=True)
    updatedAt = fields.String(required=True)


class IncidentListItemSchema(IncidentSchema):
    verificationStatus = fields.String(required=True)


class IncidentListQuerySchema(Schema):
    userId = fields.String(
        load_default=None,
        metadata={"description": "Admin only — filter by reporter UUID."},
    )
    status = fields.String(
        load_default=None,
        validate=validate.OneOf(STATUSES),
        metadata={"example": "PENDING"},
    )
    statusIn = fields.List(
        fields.String(),
        load_default=None,
        metadata={"description": "Repeat query param, e.g. statusIn=PENDING&statusIn=VERIFIED"},
    )
    urgency = fields.String(
        load_default=None,
        validate=validate.OneOf(URGENCIES),
        metadata={"example": "HIGH"},
    )
    severity = fields.String(
        load_default=None,
        validate=validate.OneOf(SEVERITIES),
    )
    type = fields.String(
        load_default=None,
        validate=validate.OneOf(INCIDENT_TYPES),
        metadata={"example": "accident"},
    )
    departmentId = fields.String(load_default=None)
    verificationStatus = fields.String(load_default=None)
    location = fields.String(load_default=None, metadata={"example": "Nairobi"})
    search = fields.String(
        load_default=None,
        metadata={"description": "Matches reference, title, location, reporter name."},
    )
    dateFrom = fields.String(
        load_default=None,
        metadata={"description": "ISO date or datetime", "example": "2026-01-01"},
    )
    dateTo = fields.String(load_default=None, metadata={"example": "2026-12-31"})
    sort = fields.String(
        load_default="urgency",
        validate=validate.OneOf(["urgency", "newest"]),
        metadata={"example": "newest"},
    )
    includeArchived = fields.Boolean(load_default=False)
    limit = fields.Integer(load_default=50, metadata={"example": 50})
    offset = fields.Integer(load_default=0, metadata={"example": 0})


class IncidentListPageSchema(Schema):
    items = fields.List(fields.Nested(IncidentListItemSchema), required=True)
    total = fields.Integer(required=True)
    limit = fields.Integer(required=True)
    offset = fields.Integer(required=True)
    hasMore = fields.Boolean(required=True)


class NoteSchema(Schema):
    id = fields.String(required=True)
    incidentId = fields.String(required=True)
    authorId = fields.String(required=True)
    authorName = fields.String(required=True)
    body = fields.String(required=True)
    createdAt = fields.String(required=True)


class MediaSchema(Schema):
    id = fields.String(required=True)
    incidentId = fields.String(required=True)
    kind = fields.String(required=True)
    url = fields.String(required=True)
    name = fields.String(required=True)
    createdAt = fields.String(required=True)


class StatusHistorySchema(Schema):
    id = fields.String(required=True)
    incidentId = fields.String(required=True)
    fromStatus = fields.String(allow_none=True)
    toStatus = fields.String(required=True)
    actorId = fields.String(required=True)
    actorName = fields.String(required=True)
    reason = fields.String(allow_none=True)
    createdAt = fields.String(required=True)


class VerificationSchema(Schema):
    id = fields.String(required=True)
    incidentId = fields.String(required=True)
    status = fields.String(required=True)
    method = fields.String(allow_none=True)
    notes = fields.String(allow_none=True)
    verifiedById = fields.String(allow_none=True)
    verifiedByName = fields.String(allow_none=True)
    verifiedAt = fields.String(allow_none=True)
    createdAt = fields.String(required=True)
    updatedAt = fields.String(required=True)


class HandoffSchema(Schema):
    id = fields.String(required=True)
    incidentId = fields.String(required=True)
    departmentId = fields.String(required=True)
    initiatedById = fields.String(required=True)
    initiatedByName = fields.String(required=True)
    status = fields.String(required=True)
    notes = fields.String(allow_none=True)
    handedOffAt = fields.String(required=True)
    acknowledgedAt = fields.String(allow_none=True)
    completedAt = fields.String(allow_none=True)
    updatedAt = fields.String(required=True)


class IncidentDetailSchema(Schema):
    """Aggregate payload for review / detail screens."""

    incident = fields.Nested(IncidentSchema, required=True)
    history = fields.List(fields.Nested(StatusHistorySchema), required=True)
    notes = fields.List(fields.Nested(NoteSchema), required=True)
    media = fields.List(fields.Nested(MediaSchema), required=True)
    verification = fields.Nested(VerificationSchema, allow_none=True)
    verifications = fields.List(fields.Nested(VerificationSchema), required=True)
    handoffs = fields.List(fields.Nested(HandoffSchema), required=True)


class MediaInputSchema(Schema):
    kind = fields.String(
        required=True,
        validate=validate.OneOf(["image", "video"]),
        metadata={"example": "image"},
    )
    url = fields.String(
        required=True,
        validate=validate.Length(min=1, max=2000),
        metadata={"example": "https://placehold.co/600x400/png"},
    )
    name = fields.String(
        required=True,
        validate=validate.Length(min=1, max=500),
        metadata={"example": "scene.png"},
    )


class CreateIncidentSchema(Schema):
    title = fields.String(
        required=True,
        validate=validate.Length(min=1, max=300),
        metadata={"example": "Multi-vehicle collision on Uhuru Highway"},
    )
    description = fields.String(
        required=True,
        validate=validate.Length(min=1, max=10000),
        metadata={
            "example": "Three vehicles involved near Bunyala Road junction. Injuries reported."
        },
    )
    type = fields.String(
        load_default="accident",
        validate=validate.OneOf(INCIDENT_TYPES),
        metadata={"example": "accident"},
    )
    urgency = fields.String(
        load_default="MEDIUM",
        validate=validate.OneOf(URGENCIES),
        metadata={"example": "HIGH"},
    )
    severity = fields.String(
        load_default="MODERATE",
        validate=validate.OneOf(SEVERITIES),
        metadata={"example": "MAJOR"},
    )
    location = fields.String(
        required=True,
        validate=validate.Length(min=1, max=500),
        metadata={"example": "Uhuru Highway, Nairobi"},
    )
    lat = fields.Float(load_default=None, allow_none=True, metadata={"example": -1.2921})
    lng = fields.Float(load_default=None, allow_none=True, metadata={"example": 36.8219})
    userId = fields.String(
        load_default=None,
        allow_none=True,
        metadata={"description": "Admin walk-in: optional reporter user UUID."},
    )
    reporterName = fields.String(
        load_default=None, allow_none=True, metadata={"example": "Walk-in caller"}
    )
    reporterEmail = fields.Email(load_default=None, allow_none=True)
    reporterPhone = fields.String(
        load_default=None, allow_none=True, metadata={"example": "+254711000000"}
    )
    preferredContactMethod = fields.String(
        load_default="PHONE",
        validate=validate.OneOf(CONTACT_METHODS),
        metadata={"example": "PHONE"},
    )
    media = fields.List(fields.Nested(MediaInputSchema), load_default=None)


class UpdateIncidentSchema(Schema):
    """Partial update — only fields present in the request body are applied. Status is not changed here."""

    title = fields.String(
        validate=validate.Length(min=1, max=300),
        metadata={"example": "Updated: collision cleared one lane"},
    )
    description = fields.String(validate=validate.Length(min=1, max=10000))
    type = fields.String(validate=validate.OneOf(INCIDENT_TYPES))
    urgency = fields.String(
        validate=validate.OneOf(URGENCIES), metadata={"example": "MEDIUM"}
    )
    severity = fields.String(validate=validate.OneOf(SEVERITIES))
    location = fields.String(validate=validate.Length(min=1, max=500))
    lat = fields.Float(allow_none=True)
    lng = fields.Float(allow_none=True)
    userId = fields.String(allow_none=True)
    reporterName = fields.String(allow_none=True)
    reporterEmail = fields.Email(allow_none=True)
    reporterPhone = fields.String(allow_none=True)
    preferredContactMethod = fields.String(validate=validate.OneOf(CONTACT_METHODS))


class ArchiveIncidentSchema(Schema):
    reason = fields.String(
        required=True,
        validate=validate.Length(min=1, max=2000),
        metadata={"example": "Duplicate of AJL-0001 — merged into primary report."},
    )


CLOSE_REASONS = [
    "FALSE_REPORT",
    "DUPLICATE",
    "UNABLE_TO_VERIFY",
    "INSUFFICIENT_INFORMATION",
    "OTHER",
]
VERIFICATION_METHODS = ["PHONE", "EMAIL", "OTHER"]
RESOLUTION_OUTCOMES = [
    "RESOLVED",
    "ASSISTANCE_PROVIDED",
    "REFERRED",
    "UNABLE_TO_ASSIST",
    "OTHER",
]


class VerifyIncidentSchema(Schema):
    method = fields.String(
        required=True,
        validate=validate.OneOf(VERIFICATION_METHODS),
        metadata={"example": "PHONE"},
    )
    notes = fields.String(
        load_default=None,
        allow_none=True,
        metadata={"example": "Caller confirmed location and casualties."},
    )


class CloseIncidentSchema(Schema):
    reason = fields.String(
        required=True,
        validate=validate.Length(min=1, max=2000),
        metadata={
            "example": "Unable to verify — no response from reporter after two attempts."
        },
    )
    reasonCode = fields.String(
        load_default="OTHER",
        validate=validate.OneOf(CLOSE_REASONS),
        metadata={"example": "UNABLE_TO_VERIFY"},
    )
    failVerification = fields.Boolean(
        load_default=False,
        metadata={"description": "Also mark latest verification as failed.", "example": True},
    )


class StartResponseSchema(Schema):
    departmentIds = fields.List(
        fields.String(),
        required=True,
        validate=validate.Length(min=1),
        metadata={
            "description": "UUIDs from GET /api/v1/departments (replace the placeholder).",
            "example": ["00000000-0000-0000-0000-000000000001"],
        },
    )
    notes = fields.String(
        load_default=None,
        allow_none=True,
        metadata={"example": "Dispatch nearest unit."},
    )


class NotifyCitizenSchema(Schema):
    sms = fields.Boolean(
        load_default=False,
        metadata={"description": "SMS dry-run until Africa's Talking is wired.", "example": False},
    )
    email = fields.Boolean(
        load_default=False,
        metadata={
            "description": "Sends via Resend to reporterEmail after commit (dry-run if no API key).",
            "example": True,
        },
    )


class ResolveIncidentSchema(Schema):
    summary = fields.String(
        required=True,
        validate=validate.Length(min=1, max=5000),
        metadata={"example": "Casualties treated on site; road cleared."},
    )
    notes = fields.String(
        load_default=None,
        allow_none=True,
        metadata={"example": "Ambulance completed handover at KNH."},
    )
    outcome = fields.String(
        required=True,
        validate=validate.OneOf(RESOLUTION_OUTCOMES),
        metadata={"example": "ASSISTANCE_PROVIDED"},
    )
    completeHandoffs = fields.Boolean(load_default=True, metadata={"example": True})
    notifyCitizen = fields.Nested(NotifyCitizenSchema, load_default=None)


class ReopenIncidentSchema(Schema):
    reason = fields.String(
        required=True,
        validate=validate.Length(min=1, max=2000),
        metadata={"example": "New information received — response needed again."},
    )


class AddNoteSchema(Schema):
    body = fields.String(
        required=True,
        validate=validate.Length(min=1, max=10000),
        metadata={"example": "Spoke with traffic police — ETA 10 minutes."},
    )


class AddMediaSchema(Schema):
    kind = fields.String(
        required=True,
        validate=validate.OneOf(["image", "video"]),
        metadata={"example": "image"},
    )
    url = fields.String(
        required=True,
        validate=validate.Length(min=1, max=2000),
        metadata={"example": "https://placehold.co/800x600/png"},
    )
    name = fields.String(
        required=True,
        validate=validate.Length(min=1, max=500),
        metadata={"example": "follow-up.png"},
    )


class UpdateHandoffSchema(Schema):
    status = fields.String(
        required=True,
        validate=validate.OneOf(
            ["PENDING", "ACKNOWLEDGED", "IN_PROGRESS", "COMPLETED", "CANCELLED"]
        ),
        metadata={"example": "ACKNOWLEDGED"},
    )
    notes = fields.String(allow_none=True, metadata={"example": "Unit en route."})


class CompleteHandoffSchema(Schema):
    notes = fields.String(
        load_default=None,
        allow_none=True,
        metadata={"example": "Patient handed over to facility."},
    )
